### ISSUE TITLE
ASP.NET Core Identity is logging users out after an hour or so

### ISSUE THREAD

**Issue by SoftCircuits** [dotnet/aspnetcore#57177]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I have just moved my application to a new server and now, when a user logs in, they are logged out after a couple of hours.

### Expected Behavior

On the old server, users would be kept log in for a considerable amount of time. If I'm not mistaken, it was 30 days.

### Steps To Reproduce

I am not doing anything unusual or out of the ordinary. This is out-of-the-box ASP.NET Core Identity. Here is my `Configure()` method.

    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        if (env.IsDevelopment())
        {
            app.UseDeveloperExceptionPage();
            app.UseMigrationsEndPoint();
        }
        else
        {
            app.UseExceptionHandler("/Error");
            app.UseStatusCodePagesWithReExecute("/ErrorCode", "?code={0}");
            // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
            app.UseHsts();
        }

        if (IsDemoSite)
        {
            app.UseHttpsRedirection();
        }
        else
        {
            RewriteOptions options = new RewriteOptions()
                .AddRedirectToHttpsPermanent()
                .AddRedirectToWwwPermanent();
            app.UseRewriter(options);
        }

        app.UseStaticFiles();

        app.UseRouting();

        app.UseAuthentication();
        app.UseAuthorization();

        app.UseEndpoints(endpoints =>
        {
            endpoints.MapControllers();
            endpoints.MapRazorPages();
        });
    }

I am testing using a production build (`env.IsDevelopment()` is false), and `IsDemoSite` is false.

### .NET Version

8.0

### Anything else?

.NET SDK:
 Version:           8.0.303
 Commit:            29ab8e3268
 Workload version:  8.0.300-manifests.34944930
 MSBuild version:   17.10.4+10fbfbf2e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.303\

.NET workloads installed:
 [wasm-tools]
   Installation Source: VS 17.10.35027.167
   Manifest Version:    8.0.7/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.workload.mono.toolchain.current\8.0.7\WorkloadManifest.json
   Install Type:        FileBased

 [aspire]
   Installation Source: VS 17.10.35027.167
   Manifest Version:    8.0.0/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.aspire\8.0.0\WorkloadManifest.json
   Install Type:        FileBased

Host:
  Version:      8.0.7
  Architecture: x64
  Commit:       2aade6beb0

.NET SDKs installed:
  8.0.303 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.28 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

**Comment by MackinnonBuck** [comment]
@SoftCircuits, have you looked at a network trace? Can you tell if cookies are being set or unset? The default should be 14 days, so this is unexpected behavior.

**Comment by SoftCircuits** [comment]
> @SoftCircuits, have you looked at a network trace? Can you tell if cookies are being set or unset? The default should be 14 days, so this is unexpected behavior.

I guess I'm not sure what I'm looking for in a network trace.

I cleared all cookies for the site and logged back in. I can now see the following two cookies.

- .AspNetCore.Antiforgery.8eba6JXrSx4
- .AspNetCore.Identity.Application

Thanks.

**Comment by halter73** [comment]
> I guess I'm not sure what I'm looking for in a network trace.

We'd want to see the Set-Cookie response header and Cookie request header. You can also look at more details like the cookie expiration after cookies are set by looking in the browser's F12 dev tools. It's under in "Application > Storage > Cookies" for Edge and Chrome. https://developer.chrome.com/docs/devtools/application/cookies/

> Also, the comment added by the auto-generated code states the default is 30 days.

Unless something else is changing it, the default is 14 days.

https://github.com/dotnet/aspnetcore/blob/eb626ad8577ac295416f645ee020f6f4024d4565/src/Security/Authentication/Cookies/src/CookieAuthenticationOptions.cs#L25-L30

The code you provided in your `public void Configure(IApplicationBuilder app, IWebHostEnvironment env)` doesn't really show what's going on. The configuration for cookies and Identity is typically in a `ConfigureServices` method if you're using a `Startup` class. Can you please provide that? [Configuring logging](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-8.0#configure-logging) to log "Debug" rather than just "Warning" logs for the "Microsoft.AspNetCore" category and looking at the logs as soon as possible after you notice your user is logged out might also help.

Unfortunately, saying that default identity logs users out after an hour or so without any additional details is not enough for us to determine what's going on. That should not happen. If this is something that you can reliably repro with a simple application, we ask that you publish the full repro app to github with step-by-step instructions to reproduce the issue and provide a link so we can take a look.

**Comment by SoftCircuits** [comment]
@halter73 The key issue here is that this was working exactly as expected. But has started logging users out after moving my site to a new server. This sort of rules out showing relevant code or providing a repro project that will reproduce the issue for everyone.

Unfortunately, I don't understand the inner workings of this enough to understand how a different server could make the difference. And I was hoping for suggestions on what I might try.

Below is my `ConfigureServices()` method. But again, this code was working just fine on the old server.

        public void ConfigureServices(IServiceCollection services)
        {
            AppSettings appSettings = Configuration.GetSection("AppSettings").Get<AppSettings>();
            IsDemoSite = appSettings.IsDemoSite;
            string connectionStringName = IsDemoSite ? "DemoConnection" : "MainConnection";

            services.AddDbContext<ApplicationDbContext>(options =>
            {
                var builder = options.UseSqlServer(
                    Configuration.GetConnectionString(connectionStringName),
                    builder => builder.MigrationsAssembly("RailtraxEntities"));
                // Enable SQL query to output window (requires appsetting setting)
                //builder.EnableSensitiveDataLogging();
            });

            services.AddDatabaseDeveloperPageExceptionFilter();

            services.AddDefaultIdentity<ApplicationUser>(options =>
                {
                    options.SignIn.RequireConfirmedAccount = true;
                })
                .AddRoles<IdentityRole>()
                .AddEntityFrameworkStores<ApplicationDbContext>();

            // Create authorization policies
            services.AddAuthorization(options =>
            {
                //// Make default page behavior to require authentication
                //options.FallbackPolicy = new AuthorizationPolicyBuilder()
                //    .RequireAuthenticatedUser()
                //    .Build();
                options.AddPolicy(Policy.Staff, p => p.RequireRole(Role.Staff));
                options.AddPolicy(Policy.Admin, p => p.RequireRole(Role.Admin));
            });

            // Set authorizations
            services.AddRazorPages(options =>
            {
                // IMPORTANT: It appears there is no method to set the default requirement for
                // all areas. When any new areas are added, we need to add authorization here.
                // Otherwise, it would be open to the public.

                options.Conventions.AuthorizeFolder("/", Policy.Staff);
                options.Conventions.AuthorizeAreaFolder("Admin", "/", Policy.Admin);
                options.Conventions.AuthorizeAreaFolder("Leasing", "/", Policy.Staff);
                options.Conventions.AuthorizeAreaFolder("Repair", "/", Policy.Staff);
                options.Conventions.AuthorizeAreaFolder("Storage", "/", Policy.Staff);
                options.Conventions.AuthorizeAreaFolder("Transloading", "/", Policy.Staff);

                options.Conventions.AllowAnonymousToPage("/Index");
                options.Conventions.AllowAnonymousToPage("/Error");
            });

            services.UseTrackAndTrace<TrackAndTraceClient>(Configuration, "TrackAndTrace");
            services.ConfigureRailInc(Configuration);
            services.AddTransient<IEmailSender, EmailSender>();
            services.AddTransient<IUserContext, RailtraxUserContext>();

            services.Configure<AppSettings>(Configuration.GetSection("AppSettings"));
            services.Configure<EmailSettings>(Configuration.GetSection("EmailSettings"));

            // Railtrax API
            services.AddControllers();
            services.AddAuthentication()
                .AddScheme<AuthenticationSchemeOptions, BasicAuthenticationHandler>("BasicAuthentication", options => { });
            services.AddAuthorization(options =>
            {
                AuthorizationPolicy policy = new AuthorizationPolicyBuilder("BasicAuthentication")
                    .RequireAuthenticatedUser()
                    .Build();
                options.AddPolicy("BasicAuthentication", policy);
            });
            services.AddScoped<IBasicAuthenticatorService, BasicAuthenticatorService>();
        }

**Comment by martincostello** [comment]
Are you storing your data protection keys in persistent storage? I notice you say you moved server.

If the keys weren't stored somewhere persistent, like blob storage, then when you moved to the new server, it wouldn't have been able to decrypt existing cookies so would have effectively logged everyone out.

There's more details about that here: https://learn.microsoft.com/aspnet/core/host-and-deploy/web-farm

**Comment by SoftCircuits** [comment]
@martincostello Not certain what you mean by protection keys. User data is stored in SQL Server.

However, the database did not move. Only the website.

Thanks.

**Comment by martincostello** [comment]
Data Protection keys are used to encrypt the contents of the cookies used for authentication, and is separate to user data. By default they're stored in either TEMP or the Windows Registry.

If you don't configure them to be stored in a stable location, then when you moved the server if you didn't move the keys too, the new server won't have been able to recognise users' existing authentication cookies as it wouldn't have the keys to decrypt them.

**Comment by SoftCircuits** [comment]
@martincostello Yeah, I haven't done anything with that. I understand the new server won't recognize existing users. But once those users log into the new server, aren't new keys created?

**Comment by martincostello** [comment]
They _should_ be provided the keys are stored in the same persistent location.

How do you host your server? Is it a physical machine that stays the same between any code redeployments, or is something like a container or VM (AKS, EKS, EC2 etc.) where scaling can make individual instances go away over time?

**Comment by SoftCircuits** [comment]
@martincostello 

> They should be provided the keys are stored in the same persistent location.

Sorry, I'm not sure exactly what you are saying here. Is there something I should do, or I should try here?

The site is hosted on a Windows VPS (both the old server and the new one). We don't have a ton of traffic. There's no multiple servers, or other containers. It's pretty simple as far as hosting goes.

**Comment by martincostello** [comment]
Say for example I host my ASP.NET Core app in Azure App Service with a single instance using the default data protection configuration. The code is deployed and things start up and data protection keys are written into the TEMP folder or registry on the server. Usage is low, so there's no auto-scaling or anything.

Some time later, there's a hardware failure in Azure, so App Service seamlessly and invisibly creates a new instance, deploys my code onto it, adds it into the load balancer's rotation, then shuts down the old instance.

As the user, I've not done anything here, but the platform has changed my compute instance from underneath me to maintain availability. In doing so, it's taken my data encryption keys with it, and effectively logged out every user of the application.

Without configuring ASP.NET Core to store data protection keys somewhere persistent, such as blob storage, I'm open to this scenario happening to my users at any time.

In this scenario, the action is on my as the application developer to store my keys somewhere that is shared between any deployments of my application regardless of what hardware they happen to be running on at any specific point in time.

It might be that you're not affected by this in this for the purposes of this issue because you have a physical server somewhere in a data centre you control that won't be swapped out on a whim, but it's still something you should be aware of, as without appropriate back-ups you can restore, if the server needed to be replaced, you'd lose your keys and all your users would get logged out.

**Comment by halter73** [comment]
I agree with @martincostello that missing data protection keys are the most likely cause of this problem. My guess is something is causing the [default key storage location](https://learn.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/default-settings) (`%LOCALAPPDATA%\ASP.NET\DataProtection-Keys` on a Windows VPS) to lose keys. Can you verify that the same keys exist in the `DataProtection-Keys` folder before and after the user is signed out?

Either way, since you're already using EF and SQL server, using [PersistKeyToDbContext](https://learn.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/overview?view=aspnetcore-8.0#persistkeystodbcontext) seems like a good fit for you. That way it's far less likely you'd lose keys.

Unfortunately, unless the above is enough for you to figure out what's going on, we still don't have enough details to confidently diagnose your issue. It could also be something like the same user on a different device doing an operation that updates the [security stamp](https://learn.microsoft.com/en-us/aspnet/core/security/authentication/identity-configuration?view=aspnetcore-8.0#isecuritystampvalidator-and-signout-everywhere) (which includes most of the Identity UI management pages) causing a signout everywhere.

Once again, I highly recommend that you [configuring logging](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-8.0#configure-logging) to log "Debug" rather than just "Warning" logs for the "Microsoft.AspNetCore" category and looking at the logs as soon as possible after you notice your user is logged out. If you've already done this, can you provide the logs?

**Comment by SoftCircuits** [comment]
@halter73 Thanks for the different ideas.

### Keys

When I attempt to navigate to `%LOCALAPPDATA%\ASP.NET\DataProtection-Keys`, I get the following message.

> Windows can't find 'C:\Users\Administrator\AppData\Local\ASP.NET\DataProtection-Keys'. Check the spelling and try again.

It appears `C:\Users\Administrator\AppData\Local\ASP.NET` does not exist.

### Logs

Before we move our main application, we moved a smaller one. Unfortunately, the smaller one does not have logging. I could look into adding it if you really thought it might shed some light on this particular issue.

### PersistKeyToDbContext

I can try this. Really, I don't understand it. I don't understand the scenarios where it is necessary, and it's never been necessary before. I will look into this further.

Thanks again.

**Comment by SoftCircuits** [comment]
@dotnet-policy-bot I answered the question. You're malfunctioning.

**Comment by MackinnonBuck** [comment]
@SoftCircuits, we'll need a repro project before we can investigate this further.

**Comment by SoftCircuits** [comment]
@MackinnonBuck I addressed this already. The exact same code was working fine. I've moved it to a new server, and now it expires user sessions within an hour. That rules out coding issues or a repro project. I'm trying to better understand how this works so that I might troubleshoot it. @halter73 had some great suggestions, which I'm working through. But a couple of his suggestions I couldn't complete for reasons I posted above.

**Comment by MackinnonBuck** [comment]
@SoftCircuits We're going to close this out, but if you find more details that point to this being a framework issue, please feel free to file a new issue with those details.

One thing to double-check is that your app is running in administrator, because this would affect the location of the data protection keys. IIS app pools usually run under a different user.
